File: app/src/main/java/com/killain/organizer/packages/interactors/DataManager.java
package com.killain.organizer.packages.interactors;

import android.content.Context;

import com.killain.organizer.packages.database.AppDatabase;
import com.killain.organizer.packages.interfaces.DBComponent;

import com.killain.organizer.packages.interfaces.DaggerDBComponent;
import com.killain.organizer.packages.interfaces.SubTaskDAO;
import com.killain.organizer.packages.interfaces.TaskDAO;
import com.killain.organizer.packages.models.SubTask;
import com.killain.organizer.packages.models.Task;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

public class DataManager {

    @Inject
    public AppDatabase db;

    @Inject
    public TaskDAO taskDAO;

    @Inject
    public SubTaskDAO subTaskDAO;

    private List<Task> taskList;

    private List<SubTask> subTaskList;
    private String date;
    private Context context;

    DBComponent component;

    public DataManager(Context context, ArrayList<Task> arrayList) {
        this.taskList = arrayList;
        this.context = context;
        component = DaggerDBComponent.builder().dBInjectorModule(new DBInjectorModule(context)).build();
        component.injectInManager(this);
    }

    public ArrayList<Task> getTasksByState(boolean isCompleted, boolean isDeleted) {
        taskList = taskDAO.getAllTasksByState(isCompleted, isDeleted);
        return (ArrayList<Task>) taskList;
    }

    public void updateTask (Task task) {
        taskDAO.updateTask(task);
    }

    public void addTask(Task task) {
        taskDAO.addTask(task);
    }

    public void addSubTask(SubTask subTask) {
        subTaskDAO.addSubTask(subTask);
    }

    public void updateSubTask(SubTask subTask) {
        subTaskDAO.updateSubTask(subTask);
    }

    public void deleteTask(Task task) {
        taskDAO.deleteTask(task);
    }

    public void setDate(String date) {
        this.date = date;
    }

    public ArrayList<SubTask> getSubTasksByReference(String reference) {
        subTaskList = subTaskDAO.getSubTasksByReference(reference);
        return (ArrayList<SubTask>) subTaskList;
    }

    public ArrayList<Task> getAllTasksByDate(String date) {
        taskList = taskDAO.getAllTasksByDate(date);
        return (ArrayList<Task>) taskList;
    }
}

File: app/src/main/java/com/killain/organizer/packages/interactors/DataManager.kt
package com.killain.organizer.packages.interactors

import android.content.Context
import android.widget.Toast
import android.widget.Toast.LENGTH_SHORT
import co.metalab.asyncawait.async

import com.killain.organizer.packages.database.AppDatabase
import com.killain.organizer.packages.interfaces.DBComponent
import com.killain.organizer.packages.interfaces.DaggerDBComponent

import com.killain.organizer.packages.interfaces.SubTaskDAO
import com.killain.organizer.packages.interfaces.TaskDAO
import com.killain.organizer.packages.models.SubTask
import com.killain.organizer.packages.models.Task
import kotlinx.coroutines.*
import kotlinx.coroutines.Dispatchers.IO
import java.util.ArrayList
import javax.inject.Inject
import kotlin.NullPointerException

class DataManager(val context: Context) {

    @Inject
    lateinit var db: AppDatabase

    @Inject
    lateinit var taskDAO: TaskDAO

    @Inject
    lateinit var subTaskDAO: SubTaskDAO

    private var taskList: List<Task>? = null

    private var subTaskList: List<SubTask>? = null
    private var date: String? = null

    var component: DBComponent? = null

//    val undeletedTasks: MutableList<Task>?
//        get() {
//            taskList = taskDAO?.undeletedTasks
//            return taskList as? ArrayList<Task>?
//        }

    init {
        component = DaggerDBComponent.builder().dBInjectorModule(DBInjectorModule(context)).build()
        component?.injectInManager(this)
    }

    fun getUndeletedTasks() : MutableList<Task?>?  {
        return if (::taskDAO.isInitialized) {
            runBlocking {
                taskList = withContext(IO) {
                    taskDAO.getUndeletedTasks()
                }
//                    taskList = taskDAO.getUndeletedTasks()
//                    return taskList as MutableList<Task?>?
            }
            taskList as MutableList<Task?>?

        } else {
            throwTaskDAOException()
            null
        }
    }

    fun getTasksByState(isCompleted: Boolean, isDeleted: Boolean): ArrayList<Task>? {
        return if (::taskDAO.isInitialized) {
            runBlocking {
                taskList = withContext(IO) {
                    taskDAO.getAllTasksByState(isCompleted, isDeleted)
                }
            }
            taskList as ArrayList<Task>?
        } else {
            throwTaskDAOException()
            null
        }
    }

    fun updateTask(task: Task?) {
        if (::taskDAO.isInitialized) {
            runBlocking {
                withContext(IO) {
                    taskDAO.updateTask(task)
                }
            }
        } else {
            throwTaskDAOException()
        }
    }

    fun customUpdateTask(task: Task?, oldText: String) {
        if (::taskDAO.isInitialized) {
            runBlocking {
                withContext(IO) {
                    taskDAO.customUpdateTask(task?.task_string,
                            task?.date, task?.time,
                            task?.isCompleted, task?.isNotificationShowed,
                            task?.isDeleted, task?.isHasReference,
                            oldText)
                }
            }
        } else {
            throwTaskDAOException()
        }
    }

    fun addTask(task: Task) {
        if (::taskDAO.isInitialized) {
            runBlocking {
                withContext(IO) {
                    taskDAO.addTask(task)
                }
            }
        } else {
            throwTaskDAOException()
        }
    }

    fun addSubTask(subTask: SubTask?) {
        if (::subTaskDAO.isInitialized) {
            runBlocking {
                withContext(IO) {
                    subTaskDAO.addSubTask(subTask)
                }
            }
        } else {
            throwSubTaskDAOException()
        }
    }

    fun updateSubTask(subTask: SubTask?) {
        if (::subTaskDAO.isInitialized) {
            runBlocking {
                withContext(IO) {
                    subTaskDAO.updateSubTask(subTask)
                }
            }
        } else {
            throwSubTaskDAOException()
        }
    }

    fun deleteSubTask(subTask: SubTask?) {
        if (::subTaskDAO.isInitialized) {
            runBlocking {
                withContext(IO) {
                    subTaskDAO.deleteSubTask(subTask)
                }
            }
        } else {
            throwSubTaskDAOException()
        }
    }

    fun deleteTask(task: Task?) {
        if (::taskDAO.isInitialized) {
            runBlocking {
                withContext(IO) {
                    taskDAO.deleteTask(task)
                }
            }
        } else {
            throwTaskDAOException()
        }
    }

    fun setDate(date: String?) {
        this.date = date
    }

    fun getSubTasksByReference(reference: String): ArrayList<SubTask>? {
        return if (::subTaskDAO.isInitialized) {
                runBlocking {
                    subTaskList = withContext(IO) {
                        subTaskDAO.getSubTasksByReference(reference)
                    }
            }
            subTaskList as ArrayList<SubTask>?
        } else {
            throwSubTaskDAOException()
            null
        }
    }

    fun getAllTasksByDate(date: Long): MutableList<Task?>? {
        return if (::taskDAO.isInitialized) {
            runBlocking {
                taskList = withContext(IO) {
                    taskDAO.getAllTasksByDate(date)
                }
            }
            taskList as? MutableList<Task?>
        } else {
            throwTaskDAOException()
            null
        }
    }

    fun destroyInstance() {

    }

    private fun throwTaskDAOException() {
        Toast.makeText(context, "TaskDAO is not initialized in DataManager", LENGTH_SHORT).show()
        throw NullPointerException("TaskDAO is not initialized in DataManager")
    }

    private fun throwSubTaskDAOException() {
        Toast.makeText(context, "SubTaskDAO is not initialized in DataManager", LENGTH_SHORT).show()
        throw NullPointerException("SubTaskDAO is not initialized in DataManager")
    }
}
